crates/notion-core/src/tool/binary.rs

```diff
@@ -1,83 +1,70 @@
+        // check if the executable is a direct dependency
+        if project.has_direct_bin(&exe)? {
+            // use the full path to the file
+            let mut path_to_bin = project.local_bin_dir();
+            path_to_bin.push(&exe);
 
-                // otherwise use the user platform.
-                if let Some(ref platform) = session.user_platform()? {
-                    let image = platform.checkout(session)?;
-                    return Ok(Self::from_components(
-                        &path_to_bin.as_os_str(),
-                        params.args,
-                        &image.path()?,
-                    ));
-                }
+            // if we're in a pinned project, use the project's platform.
+            if let Some(ref platform) = project.platform() {
+                let image = platform.checkout(session)?;
+                return Ok(ToolCommand::direct(
+                    &path_to_bin.as_os_str(),
+                    args,
+                    &image.path()?,
+                ));
+            }
 
-                // if there's no user platform selected, fail.
-                throw!(ErrorDetails::NoPlatform);
+            // otherwise use the user platform.
+            if let Some(ref platform) = session.user_platform()? {
+                let image = platform.checkout(session)?;
+                return Ok(ToolCommand::direct(
+                    &path_to_bin.as_os_str(),
+                    args,
+                    &image.path()?,
+                ));
             }
-        }
 
-        // try to use the user toolchain
-        if let Some(user_tool) = session.get_user_tool(&params.executable)? {
-            return Ok(Self::from_components(
-                &user_tool.bin_path.as_os_str(),
-                params.args,
-                &user_tool.image.path()?,
-            ));
+            // if there's no user platform selected, pass through to existing PATH.
+            return ToolCommand::passthrough(&exe, args, ErrorDetails::NoPlatform);
         }
-
-        // at this point, there is no project or user toolchain
-        // the user is executing a Notion shim that doesn't have a way to execute it
-        throw!(ErrorDetails::BinaryNotFound {
-            name: params.executable.to_string_lossy().to_string(),
-        });
     }
 
-    fn from_components(exe: &OsStr, args: ArgsOs, path_var: &OsStr) -> Self {
-        Binary(command_for(exe, args, path_var))
+    // try to use the user toolchain
+    if let Some(user_tool) = session.get_user_tool(&exe)? {
+        return Ok(ToolCommand::direct(
+            &user_tool.bin_path.as_os_str(),
+            args,
+            &user_tool.image.path()?,
+        ));
     }
 
-    fn command(self) -> Command {
-        self.0
-    }
+    // at this point, there is no project or user toolchain
+    // Pass through to the existing PATH
+    ToolCommand::passthrough(
+        &exe,
+        args,
+        ErrorDetails::BinaryNotFound {
```

Comment:
If the above is changed, then this error message should be changed to indicate that there was no platform found and the binary was not found in the system.

Reply:
This doesn't make sense now given that I don't think we should fall through to the user tool above